client/cmd/dexc-desktop/app_darwin.go

```diff
@@ -0,0 +1,421 @@
+	// cleanup. See:
+	// https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428642-applicationshouldterminate?language=objc
+	addMethodToDelegate("applicationShouldTerminate:", func(s objc.Object) core.NSUInteger {
+		cancel()                 // no-op with clean rpc/web server setup
+		wg.Wait()                // no-op with clean setup and shutdown
+		shutdownCloser.Done(log) // execute defer statements
+		shutdownCloser.Success()
+
+		if dex.FileExists(dexcDesktopStateFile) {
+			err := os.Remove(dexcDesktopStateFile)
+			if err != nil {
+				// Just log the error, we shouldn't block shutdown.
+				log.Error("failed to remove dexc-desktop state file: %w", err)
+			}
+		}
+
+		return core.NSUInteger(1)
+	})
+
+	// "applicationDockMenu" method returns the app's dock menu. See:
+	// https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428564-applicationdockmenu?language=objc
+	addMethodToDelegate("applicationDockMenu:", func(_ objc.Object) objc.Object {
+		menu := cocoa.NSMenu_New()
+		windows := cocoa.NSApp().OrderedWindows()
+		winLen := windows.Count()
+		var winIndex uint64
+		for ; winIndex < winLen; winIndex++ {
+			winObj := windows.ObjectAtIndex(winIndex)
+			window := cocoa.NSWindow_fromRef(winObj)
+			item := cocoa.NSMenuItem_New()
+			item.SetTitle(window.Title())
+			// Set target to the window so the "orderFront:" selector is
+			// executed by it.
+			item.SetTarget(window)
+			item.SetAction(objc.Sel("orderFront:"))
+			menu.AddItem(item)
+		}
+
+		if winLen > 0 {
+			menu.AddItem(cocoa.NSMenuItem_Separator())
+		}
+
+		newWindowItem := cocoa.NSMenuItem_New()
+		newWindowItem.SetTitle("New Window")
+		newWindowItem.SetAction(objc.Sel("newWindow:"))
+		addMethodToDelegate("newWindow:", func(_ objc.Object) {
+			windows := cocoa.NSApp().OrderedWindows()
+			len := windows.Count()
+			if len < uint64(maxOpenWindows) {
+				createNewAppWindow(url)
+			} else {
+				// Show the last window if maxOpenWindows has been exceeded.
+				winObj := windows.ObjectAtIndex(len - 1)
+				win := cocoa.NSWindow_fromRef(winObj)
+				win.OrderFront(nil)
+			}
+		})
+
+		openLogsItem := cocoa.NSMenuItem_New()
+		openLogsItem.SetTitle("Open Logs")
+		openLogsItem.SetAction(objc.Sel("openLogs:"))
+		addMethodToDelegate("openLogs:", func(_ objc.Object) {
+			logDirURL, err := app.FilePathToURL(logDir)
+			if err != nil {
+				log.Errorf("error constructing log directory URL: %v", err)
+			} else {
+				browser.OpenURL(logDirURL)
+			}
+		})
+
+		menu.AddItem(newWindowItem)
+		menu.AddItem(openLogsItem)
+		return menu
+	})
+
+	// MacOS will always send the "windowWillClose" event when an application
+	// window is closing. See:
+	// https://developer.apple.com/documentation/appkit/nswindowdelegate/1419605-windowwillclose?language=objc
+	var noteSent bool
+	addMethodToDelegate("windowWillClose:", func(_ objc.Object) {
+		nOpenWindows--
+
+		err := clientCore.Logout()
```

Comment:
Shouldn't logout if there are still other windows open.

Reply:
oh, yup. I had to log out when we allowed only one window. Thanks for the review.